Track 2 changes for TS/JS KeyVault

jonathandturner opened the pull request:
This PR is an update of the track 2 changes we want to do to the KeyVault TS/JS API. 

* `SecretAttributes` -> `SecretManagementAttributes`

In order to break Secret into two parts -- Secret and SecretAttributes -- the type that was already called SecretAttributes is renamed. As the docs call this "secret management attributes", I went with that.

* `Secret` -> `Secret` + `SecretAttributes`

As `getSecretVersions` caused confusions in the user studies because users expected to be able to read a value when one wasn't present, we decided to split the Secret type into two parts: `SecretAttributes` which is the original Secret minus a value, and `Secret` which extends SecretAttributes with the value. Now, `getSecretVersions` iterates over SecretAttributes, and the user will get an error if they try to access the value directly.

The corresponding `updateSecret` became `updateSecretAttributes` as part of this change.

The PR will also flatten attributes for enabled, expires, and notBefore into the Secret and into Options for set and update, so that users do not have to navigate multiple layers of object to find them.

schaabs commented on sdk/keyvault/keyvault/src/secretsModels.ts:
```diff
@@ -64,15 +66,45 @@ export interface DeletedSecret extends Secret {
+ * the corresponding 'get' call for the entity (for example: `getSecret`). 
+ */
+export class EntityVersion {
+  private version: string;
```
I'm not 100% sure about making an exported type for version.  This will give the user a better idea of what is expected in the case of APIs which take the version, but it does add another type to the package, and would have to be a type either duplicated or shared across the secrets, keys and certificate packages.  It's possible that adding a code snippet to the APIs which take in a version showing this pattern would be just as effective.

jonathandturner replied:
Moving back to string for now. We can revisit later, but seems better to align with other languages and make the decision more unilaterally.

schaabs commented on sdk/keyvault/keyvault/src/secretsModels.ts:
```diff
@@ -123,7 +155,7 @@ export interface UpdateSecretOptions extends msRest.RequestOptionsBase {
    * attributes.
    */
-  attributes?: SecretAttributes;
+  attributes?: Models.SecretManagementAttributes;
```
>attributes?: Models.SecretManagementAttributes; [](start = 2, length = 47)

This should be flattened out. i.e. the writable feilds on SecretAttributes (expires, not before and enabled)  should be fields on the options bag

jonathandturner replied:
Should I also flatten from `SecretManagementAttributes` into `SecretAttributes`?  This would effectively flatten those three fields in the Secret that users get returned, instead of having an `attribute` field on the secret.

schaabs replied:
Yes we should do this.

jonathandturner replied:
It should be flattened/unflattened in Secret as well as set/update now.

schaabs commented:
> export interface GetSecretOptions extends msRest.RequestOptionsBase {

Same comment as above regarding RequestOptionsBase

---
Refers to: sdk/keyvault/keyvault/src/secretsModels.ts:173 in 66f0724. [](commit_id = 66f072451b5c4c2f48d3fce5b3dbeab94d755210, deletion_comment = False)

schaabs commented:
> export interface GetAllSecretsOptions extends msRest.RequestOptionsBase {

Same comment as above regarding RequestOptionsBase

---
Refers to: sdk/keyvault/keyvault/src/secretsModels.ts:188 in 66f0724. [](commit_id = 66f072451b5c4c2f48d3fce5b3dbeab94d755210, deletion_comment = False)

schaabs commented:
>   maxPageSize?: number;

This field should be removed, we're not going to expose this parameter to users.

---
Refers to: sdk/keyvault/keyvault/src/secretsModels.ts:193 in 66f0724. [](commit_id = 66f072451b5c4c2f48d3fce5b3dbeab94d755210, deletion_comment = False)

schaabs commented on sdk/keyvault/keyvault/src/models/index.ts:
```diff
@@ -13,6 +13,9 @@ import * as msRest from "@azure/ms-rest-js";
 export { BaseResource, CloudError };
 
+export namespace Models {
+
```
I didn't see any changes to the index.ts in the parent directory are these being exported from the package?  If so they should not be.

jonathandturner replied:
This should be fixed now.

jeremymeng commented on sdk/keyvault/keyvault/src/keyVaultClient.ts:
```diff
@@ -10,19 +10,19 @@
 
 import * as msRest from "@azure/ms-rest-js";
 import * as msRestAzure from "@azure/ms-rest-azure-js";
-import * as Models from "./models";
```
I believe `keyVaultClient.ts` and `keyVaultClientContext.ts` are from autorest code-gen.  At one point I considered moving them into `src/generated` folder but that didn't happen.  Maybe we shouldn't change them.

jonathandturner replied:
That was a mistake, I'll fix it.

jonathandturner commented:
Addressed the feedback. I went ahead and pulled out api-extractor from this PR. I'll follow-up with another PR that adds API extractor support separately.

schaabs commented:
> 

We're going to be separating secrets, keys and certificates into their own client.  This file should be moved into a subfolder secrets, and the secrets folder should have it's own index.ts

---
Refers to: sdk/keyvault/keyvault/src/secretsClient.ts:3 in 514e26e. [](commit_id = 514e26e534f4357b59da3a9a30ee30a200184c78, deletion_comment = False)

schaabs commented on sdk/keyvault/keyvault/src/secretsModels.ts:
```diff
@@ -1,14 +1,13 @@
 import * as msRest from "@azure/ms-rest-js";
-import { SecretAttributes } from "./models";
```
We're going to be separating secrets, keys and certificates into their own client.  This file should be moved into a subfolder secrets, and the secrets folder should have it's own index.ts

schaabs commented:
> export interface SetSecretOptions extends msRest.RequestOptionsBase {

This should have a RequestOptionsBase property rather than extending this interface.

---
Refers to: sdk/keyvault/keyvault/src/secretsModels.ts:97 in 514e26e. [](commit_id = 514e26e534f4357b59da3a9a30ee30a200184c78, deletion_comment = False)

schaabs commented:
> export interface UpdateSecretOptions extends msRest.RequestOptionsBase {

Same comment as above.

---
Refers to: sdk/keyvault/keyvault/src/secretsModels.ts:130 in 514e26e. [](commit_id = 514e26e534f4357b59da3a9a30ee30a200184c78, deletion_comment = False)
